elastic/synthetics-recorder
# feat: Step divider UI

## Functional improvements/tasks

I've compiled a list of things that must be addressed before we can do a second round from the reviewers.

- [x] UX around `trash` button is confusing, does it delete the separator itself or all its content? https://github.com/elastic/synthetics-recorder/commit/2b81236a44ad5938656348cc2621456e835b45f6
- [x] UX of drag/drop feels confusing, why are we dragging separators instead of moving actions individually into buckets?
  **NOTE:** I'm starting this one out as resolved, please see [this comment](https://github.com/elastic/synthetics-recorder/issues/64#issuecomment-974334797) for further clarification as to why separators are dragged vs. individual actions. Happy to chat offline or discuss further on the issue.
- [x] Renaming a step changes the placeholder https://github.com/elastic/synthetics-recorder/pull/175/commits/0d3ecbd2952574045c130b1e5cfa2d91768eea59
- [x] Extra `open` and `closePage` actions 7c47cad925f447966f42810b41c7b03a270cd957
- [x] `grab`/`grabbing` cursor displays on non-draggable elements (doesn't seem to happen for me at the moment. Only the first step is not draggable).
- [x] No tooltips on new buttons (delete button is confusing) 2b81236a44ad5938656348cc2621456e835b45f6
- [x] Drop area is too small, difficult to actually place the cursor in it to split the steps 1d351d4bfbfdb0e41f07f8c9860d630db9cda0cf
- [x] Use EUI page color for background f38838e237a099222ba92e2f5f6709ee8d7e0ac0
- [x] Do not allow steps to be deleted while collapsed 5f6ffa5d0125cd2c4220368f6ad8158c6d745ae8
- [x] Too much space between steps accec6e88533e474181811a35a9f572e81c9b644
- [x] Delete button is visible during step edit mode and clicking it results in weird behavior. Hide the button. https://github.com/elastic/synthetics-recorder/pull/175/commits/1385855a3207aec4022b61a3065ade560d2b76f7
- [x] Test result status indicator color does not work with custom-named steps 9fd1cb8f3fd491a91e7779585f4bb0a5360ebfb5

## Summary

**Note to reviewers:** at the moment I'm only expecting functional reviews. There is a lot of overlap with #171 as this is built on top of it. I want to get the UX problems ironed out first, as these PRs need to be merged more-or-less in unison, as #171 breaks the existing step functionality without including a UI method of adding/moving/removing steps. Once it is merged, we will do an in-depth review of the code.

Resolves #64.

Resolves #87 as it is the last unresolved item in that meta-ticket.

Depends on #171.

Adds drag/drop functionality for steps, and other related controls/interaction. Complete list:

- steps are drag/droppable (i.e. reorganize existing steps around the recorded actions
- steps are insertable
- step dividers are deletable
- steps have an assignable name, that can be edited via the UI
- steps remain collapsible



## Implementation details

There are a few deviations from the design as it is shown in the issue:
- I did not add a drag/drop icon as the design shows; my thinking is that the functionality is discoverable due to the icon changing when the user passes the cursor over a divider.
- There was no design specification in the issue deliverable for how the droppable areas should look. I've emulated a bit of the EUI drag/drop UI from one of the examples, using the primary and secondary colors from the EUI palette, but this is completely open for debate/improvement.

## How to validate this change

Test the functionality of each of the items related in the description.
- Add a new step divider between actions
- You should be able to drag dividers between the actions for the preceding and next step.
- Deleting a divider should merge its nested steps to the preceding one
- The first step should not be draggable or deletable
- Rename a step, ensure you see it reflected in the code output
- Get rid of the custom name (i.e., save empty string), the original default name should reappear
- Test the renaming in combination with dragging, deleting, creating steps

Refer to the issue for additional input into the functionality.
